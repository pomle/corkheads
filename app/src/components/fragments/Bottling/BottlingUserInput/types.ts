export type FieldName =
  | "abv"
  | "age"
  | "bottleCode"
  | "bottleCount"
  | "bottleLabel"
  | "bottlerName"
  | "bottlerCountry"
  | "bottleSize"
  | "bottlingDate"
  | "bottlingYear"
  | "category"
  | "displayName"
  | "distillBatchNo"
  | "distillCaskNo"
  | "distillCaskType"
  | "distillDate"
  | "distillYear"
  | "distillerCountry"
  | "distillerDistrict"
  | "distillerName"
  | "series";

export type Fields<T> = Record<FieldName, T>;

export type Entries = Fields<string>;
