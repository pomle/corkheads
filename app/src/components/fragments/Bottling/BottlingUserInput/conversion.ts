import { asABV } from "lib/format/stringify";
import moment from "moment";
import { Moment } from "moment";
import { Bottling, createBottling } from "types/Bottling";
import { Entries } from "./types";

export function toEntries(bottling: Bottling): Entries {
  const { distill } = bottling;
  const { distillery } = distill;

  return {
    abv: distill.alcoholByVolumePercentage
      ? asABV(distill.alcoholByVolumePercentage)
      : "",
    age: distill.age?.toString() || "",
    bottleCode: bottling.code || "",
    bottleCount: bottling.bottlesProduced?.toString() || "",
    bottleLabel: bottling.label || "",
    bottlerName: bottling.bottler.name || "",
    bottlerCountry: bottling.bottler.country || "",
    bottleSize: bottling.bottleSize || "",
    bottlingDate: dateToString(bottling.date),
    bottlingYear: bottling.year?.toString() || "",
    distillBatchNo: distill.batch.number || "",
    distillCaskNo: distill.cask.number || "",
    distillCaskType: distill.cask.type || "",
    distillDate: dateToString(distill.date),
    distillYear: distill.year?.toString() || "",
    distillerCountry: distillery.country || "",
    distillerDistrict: distillery.district || "",
    distillerName: distillery.name || "",
    series: bottling.series.name || "",
  };
}

export function toBottling(entries: Entries): Bottling {
  const bottling = createBottling();

  const alcoholPercentage = fromPercentInput(entries.abv);
  if (!isNaN(alcoholPercentage)) {
    bottling.distill.alcoholByVolumePercentage = alcoholPercentage;
  }

  const agedTimeYears = toNumberMaybe(entries.age);
  if (!isNaN(agedTimeYears)) {
    bottling.distill.age = agedTimeYears;
  }

  const bottleCode = entries.bottleCode;
  if (bottleCode.length) {
    bottling.code = bottleCode;
  }

  const bottleCount = toNumberMaybe(entries.bottleCount);
  if (!isNaN(bottleCount)) {
    bottling.bottlesProduced = bottleCount;
  }

  const bottleLabel = entries.bottleLabel;
  if (bottleLabel.length) {
    bottling.label = bottleLabel;
  }

  const bottleSize = entries.bottleSize;
  if (bottleSize.length) {
    bottling.bottleSize = bottleSize;
  }

  const bottlingYear = toNumberMaybe(entries.bottlingYear);
  if (!isNaN(bottlingYear)) {
    bottling.year = bottlingYear;
  }

  const bottlingDate = toDateMaybe(entries.bottlingDate);
  if (bottlingDate) {
    bottling.date = bottlingDate;
  }

  const bottlerName = entries.bottlerName;
  if (bottlerName.length > 0) {
    bottling.bottler.name = bottlerName;
  }

  const bottlerCountry = entries.bottlerCountry;
  if (bottlerCountry.length > 0) {
    bottling.bottler.country = bottlerCountry;
  }

  const distillerName = entries.distillerName;
  if (distillerName.length > 0) {
    bottling.distill.distillery.name = distillerName;
  }

  const distillerCountry = entries.distillerCountry;
  if (distillerCountry.length > 0) {
    bottling.distill.distillery.country = distillerCountry;
  }

  const distillerDistrict = entries.distillerDistrict;
  if (distillerDistrict.length > 0) {
    bottling.distill.distillery.district = distillerDistrict;
  }

  const distillYear = toNumberMaybe(entries.distillYear);
  if (distillYear) {
    bottling.distill.year = distillYear;
  }

  const distillDate = toDateMaybe(entries.distillDate);
  if (distillDate) {
    bottling.distill.date = distillDate;
  }

  const distillCaskNo = entries.distillCaskNo;
  if (distillCaskNo.length > 0) {
    bottling.distill.cask.number = distillCaskNo;
  }

  const distillCaskType = entries.distillCaskType;
  if (distillCaskType.length > 0) {
    bottling.distill.cask.type = distillCaskType;
  }

  const distillBatchNo = entries.distillBatchNo;
  if (distillBatchNo.length > 0) {
    bottling.distill.batch.number = distillBatchNo;
  }

  const seriesName = entries.series;
  if (seriesName.length > 0) {
    bottling.series.name = seriesName;
  }

  return bottling;
}

function toNumberMaybe(source: unknown) {
  const maybeNumber = parseFloat(source as string);
  if (isFinite(maybeNumber)) {
    return maybeNumber as number;
  }
  return NaN;
}

function toDateMaybe(source: unknown) {
  const maybeDate = moment(source as string);
  if (maybeDate.isValid()) {
    return maybeDate;
  }
  return undefined;
}

function fromPercentInput(source: unknown) {
  if (typeof source === "string") {
    const text = source.replace(/[^0-9,.]/g, "");
    const decimalized = text.replace(/[,]/g, ".");
    return toNumberMaybe(decimalized);
  }
  return NaN;
}

function dateToString(date?: Moment) {
  if (date) {
    if (date.isValid()) {
      return date.toISOString();
    }
  }
  return "";
}
