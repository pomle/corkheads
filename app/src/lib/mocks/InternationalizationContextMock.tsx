import React, { useMemo, useState } from "react";
import { Locale } from "lib/i18n/localization";
import { Context } from "components/context/InternationalizationContext";

interface InternationalizationContextMockProps {
  initialLocale?: Locale;
}

export const InternationalizationContextMock: React.FC<InternationalizationContextMockProps> = ({
  initialLocale = Locale.enGB,
  children,
}) => {
  const locale = useState<Locale>(initialLocale);

  const value = useMemo(() => {
    return {
      locale,
    };
  }, [locale]);

  return <Context.Provider value={value}>{children}</Context.Provider>;
};
