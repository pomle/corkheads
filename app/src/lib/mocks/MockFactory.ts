import moment from "moment";
import { Bottling } from "types/Bottling";

export function createMockFactory() {
  function createBottling(): Bottling {
    return {
      barcodes: [],
      bottlesProduced: 2033,
      bottleSize: "75cl",
      bottler: {
        name: "Bottlerama",
        country: "Bottleville",
      },
      code: "L-214LA23",
      date: moment("2002-01-03T16:23:32Z"),
      distill: {
        alcoholByVolumePercentage: 48.23241241,
        age: 16,
        batch: {
          number: "Batch No. 123",
        },
        cask: {
          number: "Cask No. 6125",
          type: "Olorosso",
        },
        date: moment("1998-11-23T22:01:11Z"),
        distillery: {
          name: "Brukveda",
          country: "Sweden",
        },
        year: 1724,
      },
      label: "Old Marks",
      series: {
        name: "Serial Drink",
      },
      year: 1923,
    };
  }

  return {
    createBottling,
  };
}
