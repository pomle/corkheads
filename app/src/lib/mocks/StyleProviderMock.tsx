import React from "react";
import StylesProvider from "@material-ui/styles/StylesProvider";

export const StyleProviderMock: React.FC = ({ children }) => {
  return (
    <StylesProvider generateClassName={(rule) => rule.key}>
      {children}
    </StylesProvider>
  );
};
