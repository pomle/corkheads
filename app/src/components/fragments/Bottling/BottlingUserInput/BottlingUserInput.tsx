import React, { useMemo } from "react";
import { makeStyles } from "@material-ui/styles";
import EntryList from "components/ui/layout/EntryList";
import Entry from "components/ui/layout/Entry";
import { useUserInput } from "components/hooks/useUserInput";
import { Bottling } from "types/Bottling";
import { toBottling, toEntries } from "./conversion";

const useStyles = makeStyles({
  BottlingUserInput: {
    "& section h4": {
      margin: "32px 0 8px 0",
    },
  },
});

interface BottlingUserInputProps {
  bottling: Bottling;
  onChange: (bottling: Bottling) => void;
}

const BottlingUserInput: React.FC<BottlingUserInputProps> = ({
  bottling,
  onChange,
}) => {
  const entries = useMemo(() => {
    return toEntries(bottling);
  }, [bottling]);

  const userInput = useUserInput(entries, (entries) => {
    onChange(toBottling(entries));
  });

  const classes = useStyles();

  return (
    <div className={classes.BottlingUserInput}>
      <section>
        <EntryList>
          <Entry name="Label">
            <input
              type="text"
              placeholder="Label name"
              {...userInput.bottleLabel}
            />
          </Entry>

          <Entry name="Series">
            <input
              type="text"
              placeholder="Series name"
              {...userInput.series}
            />
          </Entry>

          <Entry name="Distillery">
            <input
              type="text"
              placeholder="Name of distiller"
              {...userInput.distillerName}
            />
          </Entry>

          <Entry name="Bottled by">
            <input
              type="text"
              placeholder="Name of bottler"
              {...userInput.bottlerName}
            />
          </Entry>

          <Entry name="Age">
            <input type="number" placeholder="# of years" {...userInput.age} />
          </Entry>

          <Entry name="Alcohol">
            <input type="number" placeholder="%" {...userInput.abv} />
          </Entry>
        </EntryList>
      </section>

      <section>
        <h4>The Whisky</h4>

        <EntryList>
          <Entry name="Date distilled">
            <input
              type="number"
              placeholder="Year"
              {...userInput.distillYear}
            />
          </Entry>

          <Entry name="Date bottled">
            <input
              type="number"
              placeholder="Year"
              {...userInput.bottlingYear}
            />
          </Entry>

          <Entry name="Age">
            <input type="number" placeholder="# of years" {...userInput.age} />
          </Entry>

          <Entry name="Cask No.">
            <input
              type="text"
              placeholder="Cask serial no."
              {...userInput.distillCaskNo}
            />
          </Entry>

          <Entry name="Cask type">
            <input
              type="text"
              placeholder="Olorosso, Bourbon"
              {...userInput.distillCaskType}
            />
          </Entry>

          <Entry name="Batch No.">
            <input
              type="text"
              placeholder="Batch serial no."
              {...userInput.distillBatchNo}
            />
          </Entry>

          <Entry name="Alcohol">
            <input type="number" placeholder="%" {...userInput.abv} />
          </Entry>
        </EntryList>
      </section>

      <section>
        <h4>The Distillery</h4>
        <EntryList>
          <Entry name="Distilled by">
            <input
              type="text"
              placeholder="Distillery name"
              {...userInput.distillerName}
            />
          </Entry>

          <Entry name="Country">
            <input
              type="text"
              placeholder="ex. Scotland"
              {...userInput.distillerCountry}
            />
          </Entry>

          <Entry name="District">
            <input
              type="text"
              placeholder="ex. Islay"
              {...userInput.distillerDistrict}
            />
          </Entry>
        </EntryList>
      </section>

      <section>
        <h4>The Bottle</h4>

        <EntryList>
          <Entry name="Bottled by">
            <input
              type="text"
              placeholder="Name of bottler"
              {...userInput.bottlerName}
            />
          </Entry>

          <Entry name="Bottles produced">
            <input
              type="number"
              placeholder="# of bottles"
              {...userInput.bottleCount}
            />
          </Entry>

          <Entry name="Bottle sizes">
            <input
              type="text"
              placeholder="ex. 70cl, 40cl"
              {...userInput.bottleSize}
            />
          </Entry>

          <Entry name="Bottle code">
            <input
              type="text"
              placeholder="ex. L6 3224"
              {...userInput.bottleCode}
            />
          </Entry>

          <Entry name="Barcode">
            <input type="number" placeholder="ex. 312541512" />
          </Entry>
        </EntryList>
      </section>
    </div>
  );
};

export default BottlingUserInput;
