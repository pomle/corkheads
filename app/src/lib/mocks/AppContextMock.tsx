import React from "react";
import { StyleProviderMock } from "./StyleProviderMock";
import Themer from "components/ui/theme/Themer";

export const AppContextMock: React.FC = ({ children }) => {
  return (
    <StyleProviderMock>
      <Themer theme="dusk">{children}</Themer>
    </StyleProviderMock>
  );
};
