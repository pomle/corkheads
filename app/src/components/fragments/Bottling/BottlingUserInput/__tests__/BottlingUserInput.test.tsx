import React from "react";
import renderer from "react-test-renderer";
import { AppContextMock } from "lib/mocks/AppContextMock";
import { createMockFactory } from "lib/mocks/MockFactory";
import BottlingUserInput from "../BottlingUserInput";

const mocks = createMockFactory();

describe("BottlingUserInput component", () => {
  it("renders as expected", () => {
    const bottling = mocks.createBottling();
    const onChange = jest.fn();
    const component = renderer.create(
      <AppContextMock>
        <BottlingUserInput bottling={bottling} onChange={onChange} />
      </AppContextMock>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
