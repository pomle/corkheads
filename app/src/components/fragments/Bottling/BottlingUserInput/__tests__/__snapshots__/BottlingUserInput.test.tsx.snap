// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BottlingUserInput component renders as expected 1`] = `
<div
  className="BottlingUserInput"
>
  <section>
    <div
      className="EntryList"
    >
      <dl
        className="Entry"
      >
        <dt>
          Label
        </dt>
        <dd>
          <input
            onChange={[Function]}
            placeholder="Label name"
            type="text"
            value="Old Marks"
          />
        </dd>
      </dl>
      <div
        className="Divider"
      />
      <dl
        className="Entry"
      >
        <dt>
          Series
        </dt>
        <dd>
          <input
            onChange={[Function]}
            placeholder="Series name"
            type="text"
            value="Serial Drink"
          />
        </dd>
      </dl>
      <div
        className="Divider"
      />
      <dl
        className="Entry"
      >
        <dt>
          Distillery
        </dt>
        <dd>
          <input
            onChange={[Function]}
            placeholder="Name of distiller"
            type="text"
            value="Brukveda"
          />
        </dd>
      </dl>
      <div
        className="Divider"
      />
      <dl
        className="Entry"
      >
        <dt>
          Bottled by
        </dt>
        <dd>
          <input
            onChange={[Function]}
            placeholder="Name of bottler"
            type="text"
            value="Bottlerama"
          />
        </dd>
      </dl>
      <div
        className="Divider"
      />
      <dl
        className="Entry"
      >
        <dt>
          Age
        </dt>
        <dd>
          <input
            onChange={[Function]}
            placeholder="# of years"
            type="number"
            value="16"
          />
        </dd>
      </dl>
      <div
        className="Divider"
      />
      <dl
        className="Entry"
      >
        <dt>
          Alcohol
        </dt>
        <dd>
          <input
            onChange={[Function]}
            placeholder="%"
            type="text"
            value="48.2%"
          />
        </dd>
      </dl>
    </div>
  </section>
  <section>
    <h4>
      The Whisky
    </h4>
    <div
      className="EntryList"
    >
      <dl
        className="Entry"
      >
        <dt>
          Date distilled
        </dt>
        <dd>
          <input
            onChange={[Function]}
            placeholder="Year"
            type="number"
            value="1724"
          />
        </dd>
      </dl>
      <div
        className="Divider"
      />
      <dl
        className="Entry"
      >
        <dt>
          Date bottled
        </dt>
        <dd>
          <input
            onChange={[Function]}
            placeholder="Year"
            type="number"
            value="1923"
          />
        </dd>
      </dl>
      <div
        className="Divider"
      />
      <dl
        className="Entry"
      >
        <dt>
          Age
        </dt>
        <dd>
          <input
            onChange={[Function]}
            placeholder="# of years"
            type="number"
            value="16"
          />
        </dd>
      </dl>
      <div
        className="Divider"
      />
      <dl
        className="Entry"
      >
        <dt>
          Cask No.
        </dt>
        <dd>
          <input
            onChange={[Function]}
            placeholder="Cask serial no."
            type="text"
            value="Cask No. 6125"
          />
        </dd>
      </dl>
      <div
        className="Divider"
      />
      <dl
        className="Entry"
      >
        <dt>
          Cask type
        </dt>
        <dd>
          <input
            onChange={[Function]}
            placeholder="Olorosso, Bourbon"
            type="text"
            value="Olorosso"
          />
        </dd>
      </dl>
      <div
        className="Divider"
      />
      <dl
        className="Entry"
      >
        <dt>
          Batch No.
        </dt>
        <dd>
          <input
            onChange={[Function]}
            placeholder="Batch serial no."
            type="text"
            value="Batch No. 123"
          />
        </dd>
      </dl>
      <div
        className="Divider"
      />
      <dl
        className="Entry"
      >
        <dt>
          Alcohol
        </dt>
        <dd>
          <input
            onChange={[Function]}
            placeholder="%"
            type="text"
            value="48.2%"
          />
        </dd>
      </dl>
    </div>
  </section>
  <section>
    <h4>
      The Distillery
    </h4>
    <div
      className="EntryList"
    >
      <dl
        className="Entry"
      >
        <dt>
          Distilled by
        </dt>
        <dd>
          <input
            onChange={[Function]}
            placeholder="Distillery name"
            type="text"
            value="Brukveda"
          />
        </dd>
      </dl>
      <div
        className="Divider"
      />
      <dl
        className="Entry"
      >
        <dt>
          Country
        </dt>
        <dd>
          <input
            onChange={[Function]}
            placeholder="ex. Scotland"
            type="text"
            value="Sweden"
          />
        </dd>
      </dl>
      <div
        className="Divider"
      />
      <dl
        className="Entry"
      >
        <dt>
          District
        </dt>
        <dd>
          <input
            onChange={[Function]}
            placeholder="ex. Islay"
            type="text"
            value=""
          />
        </dd>
      </dl>
    </div>
  </section>
  <section>
    <h4>
      The Bottle
    </h4>
    <div
      className="EntryList"
    >
      <dl
        className="Entry"
      >
        <dt>
          Bottled by
        </dt>
        <dd>
          <input
            onChange={[Function]}
            placeholder="Name of bottler"
            type="text"
            value="Bottlerama"
          />
        </dd>
      </dl>
      <div
        className="Divider"
      />
      <dl
        className="Entry"
      >
        <dt>
          Bottles produced
        </dt>
        <dd>
          <input
            onChange={[Function]}
            placeholder="# of bottles"
            type="number"
            value="2033"
          />
        </dd>
      </dl>
      <div
        className="Divider"
      />
      <dl
        className="Entry"
      >
        <dt>
          Bottle sizes
        </dt>
        <dd>
          <input
            onChange={[Function]}
            placeholder="ex. 70cl, 40cl"
            type="text"
            value="75cl"
          />
        </dd>
      </dl>
      <div
        className="Divider"
      />
      <dl
        className="Entry"
      >
        <dt>
          Bottle code
        </dt>
        <dd>
          <input
            onChange={[Function]}
            placeholder="ex. L6 3224"
            type="text"
            value="L-214LA23"
          />
        </dd>
      </dl>
      <div
        className="Divider"
      />
      <dl
        className="Entry"
      >
        <dt>
          Barcode
        </dt>
        <dd>
          <input
            placeholder="ex. 312541512"
            type="number"
          />
        </dd>
      </dl>
    </div>
  </section>
</div>
`;
